src/erlcloud_ddb_impl.erl

```diff
@@ -99,6 +107,12 @@ request(Config0, Operation, Json) ->
 %% This algorithm is similar, except that it waits a random interval up to 2^(Attempt-2)*100ms. The average
 %% wait time should be the same as boto.
 
+-spec maybe_request_and_retry(aws_config(), headers(), jsx:json_text(), {attempt, non_neg_integer()}, boolean()) -> json_return().
+maybe_request_and_retry(Config, Headers, Body, Attempt, false) ->
+    request_and_retry(Config, Headers, Body, Attempt);
+maybe_request_and_retry(_Config, Headers, Body, _Attempt, true) ->
+    {ok, {request, Headers, Body}}.
```

Comment:
Question here—are you sure you want encoded JSON, or would it be better to return the `jsx:json_term()` here? Not a suggested change, just wondering if it would be more useful to get the "raw" structure here.

Reply:
Yes, it makes more sense to return the raw structure rather than the encoded body, also I don't think the headers are important to me.

For this situation I could create the headers and encode the body inside `maybe_request_and_retry` and return the raw structure in case `no_request` is set to true or, the easier way, I let it be the way it is and return `{ok, {request, {encoded, Headers, Body}, {raw, Json}}}`.

What do you think? For me, it is indifferent since I will only use the `{raw, Json}`.

Reply:
Even if they're not important "to you" (and since this is a public API), do you imagine they could be useful for somebody else? If so, I'd ease the consumption for the "others" and keep as much information available, as possible, as long as it makes sense. @nalundgaard?

Reply:
One thing that might be handy here (and consistent with the erlcloud_ddb design) is to use a record, e.g.

```erlang
-record(ddb_request, {
    headers = [] :: headers(),
    body = <<>> :: iodata(),
    json :: jsx:json_term() | undefined
}).
```

If you write this into `erlcloud_ddb2.hrl`, you should be able to use it all over, and let clients use it. This will let us compatibly extend what is returned, if we ever care, and also make it succinct for clients to match out the parts of the record they care about. Another option, which would be what I'd do for a "new" API but not for _this one_, is use a map:

```erlang
-type ddb_request() :: #{headers := headers(), body := iodata(), json := jsx:json_term()}.
```

I prefer maps in general, but records is very much the style for this API, and I have to say that records are a lot less ambiguous 'out of the box' (because they have record tags). You can solve that by putting a property into the map (like `#{type => ddb_request}`), but anyway—I'd use a record here.

Reply:
This now returns a record containing headers, body, and JSON.
From my point of view, this is a better approach since it doesn't break the coding style of this library and clients can take advantage of it.